pkg/bundle/source.go

```diff
@@ -369,12 +387,21 @@ func deduplicateBundles(bundles []string) ([]string, error) {
 			// check existence of the hash
 			if _, ok := certificatesHashes[hash]; !ok {
 				// neew to trim a newline which is added by Encoder
-				dedupCerts = append(dedupCerts, string(bytes.Trim(pem.EncodeToMemory(block), "\n")))
-				certificatesHashes[hash] = struct{}{}
+				certificatesHashes[hash] = string(bytes.Trim(pem.EncodeToMemory(block), "\n"))
 			}
 		}
+	}
+
+	var orderedKeys [][32]byte
+	for key := range certificatesHashes {
+		orderedKeys = append(orderedKeys, key)
+	}
+	sort.Sort(ByHash(orderedKeys))
```

Comment:
Wouldn't it be simpler to just sort the slice now named `sortedDeduplicatedCerts` after it's constructed? This will avoid the custom `ByHash` sorter, as we can just use https://pkg.go.dev/sort#Strings. We don't require a particular order, it just has to be consistent.

Reply:
That would potentially move more memory around. This way we keep the amount of memcopy to a minimum. We would also have to rehash for every compare. My code is inspired by the golang book on how to iterate a hashmap in a deterministic order. Personally I do not like those custom sorters either but that is how go imagines how this should work.

I can change it but it will probably perform worse.

Reply:
Ah I think I now understand what you mean. You want to sort the buffer without hashing. Just the strings. Guess that might work as well. Since we hash anyway for deduplication the current way is quite cheap. But yeah I see.

Reply:
This would be the alternative implementation:

```
// remove duplicate certificates from bundles and sort certificates by hash
func deduplicateAndSortBundles(bundles []string) ([]string, error) {
	var block *pem.Block

	var certificatesHashes = make(map[[32]byte]struct{})
	var dedupCerts []string

	for _, cert := range bundles {
		certBytes := []byte(cert)

	LOOP:
		for {
			block, certBytes = pem.Decode(certBytes)
			if block == nil {
				break LOOP
			}

			if block.Type != "CERTIFICATE" {
				return nil, fmt.Errorf("couldn't decode PEM block containing certificate")
			}

			// calculate hash sum of the given certificate
			hash := sha256.Sum256(block.Bytes)
			// check existence of the hash
			if _, ok := certificatesHashes[hash]; !ok {
				// neew to trim a newline which is added by Encoder
				dedupCerts = append(dedupCerts, string(bytes.Trim(pem.EncodeToMemory(block), "\n")))
				certificatesHashes[hash] = struct{}{}
			}
		}

	}

	sort.Strings(dedupCerts)

	return dedupCerts, nil
}
```

Interestingly, the order is the same as with the hashed version. It breaks two tests though and I do not understand exactly why.

Reply:
I think it's worth it - for the sake of readability. But I'll leave it up to maintainers to decide. 😉

Reply:
WDYT @inteon?

Reply:
Can we keep the current code and use `slices.SortFunc(result, bytes.Compare)`?

Reply:
see https://github.com/cert-manager/trust-manager/pull/380/commits/a61c915f7aed998df005700e68c6779e73ba5ebb

Reply:
nice solution!